File: 2818_apply_operations_to_maximize_score.cpp
class Solution {
public:
    int maximumScore(vector<int>& a, int k) {
        int m = 1e9+7, n = a.size();
        vector<int> f(1e5+2), g(n), l(n, -1), r(n, n);
        iota(f.begin(), f.end(), 0);
        for (int i = 2; i * i < f.size(); ++i)
            for (int j = i*i; j < f.size(); j += i)
                f[j] = min(f[j], i);
        for (int i = 0; i < n; ++i) {
            unordered_set<int> s;
            for (int x = a[i]; x > 1; x /= f[x]) s.insert(f[x]);
            g[i] = s.size();
        }
        stack<int> z;
        for (int i = n-1; i >= 0; --i) {
            while (!z.empty() && g[z.top()] <= g[i]) l[z.top()] = i, z.pop();
            z.push(i);
        }
        z = {};
        for (int i = 0; i < n; ++i) {
            while (!z.empty() && g[z.top()] < g[i]) r[z.top()] = i, z.pop();
            z.push(i);
        }
        vector<int> idx(n);
        iota(idx.begin(), idx.end(), 0);
        sort(idx.begin(), idx.end(), [&](int i, int j){
            return a[i] > a[j] || (a[i] == a[j] && i < j);
        });
        long res = 1;
        for (int i : idx) {
            long cnt = (long)(i - l[i]) * (r[i] - i);
            cnt = min((long)k, cnt); k -= cnt;
            long p = 1, b = a[i];
            while (cnt) {
                if (cnt & 1) p = p * b % m;
                b = b * b % m;
                cnt >>= 1;
            }
            res = res * p % m;
            if (!k) break;
        }
        return res;
    }
};

File: 2818_apply_operations_to_maximize_score.py
class Solution(object):
    def maximumScore(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        import sys
        sys.setrecursionlimit(10**7)
        from collections import defaultdict
        from functools import cmp_to_key

        MOD = 10**9 + 7
        n = len(nums)
        MAX = 100002

        # Step 1: Smallest Prime Factor (SPF) Sieve
        f = list(range(MAX))
        for i in range(2, int(MAX**0.5) + 1):
            for j in range(i*i, MAX, i):
                f[j] = min(f[j], i)

        # Step 2: Count distinct prime factors for each nums[i]
        g = [0] * n
        for i in range(n):
            s = set()
            x = nums[i]
            while x > 1:
                s.add(f[x])
                x //= f[x]
            g[i] = len(s)

        # Step 3: Left and Right bounds (monotonic stack)
        l, r = [-1] * n, [n] * n
        stack = []
        for i in range(n - 1, -1, -1):
            while stack and g[stack[-1]] <= g[i]:
                l[stack.pop()] = i
            stack.append(i)

        stack = []
        for i in range(n):
            while stack and g[stack[-1]] < g[i]:
                r[stack.pop()] = i
            stack.append(i)

        # Step 4: Sort indices by nums[i] descending, then i ascending
        idx = list(range(n))
        idx.sort(key=lambda i: (-nums[i], i))

        # Step 5: Compute result using modular exponentiation
        def mod_pow(base, exp, mod):
            res = 1
            while exp:
                if exp % 2:
                    res = res * base % mod
                base = base * base % mod
                exp //= 2
            return res

        res = 1
        for i in idx:
            cnt = (i - l[i]) * (r[i] - i)
            cnt = min(k, cnt)
            k -= cnt
            res = res * mod_pow(nums[i], cnt, MOD) % MOD
            if k == 0:
                break

        return res
